acl/acl.go

```diff
@@ -328,6 +328,34 @@ type PolicyACL struct {
 	operatorRule string
 }
 
+func enforce(rule string, requiredPermission string) (allow, recurse bool) {
+	switch rule {
+	case PolicyWrite:
+		// grants read, list and write permissions
+		return true, false
+	case PolicyList:
+		// grants read and list permissions
+		if requiredPermission == PolicyList || requiredPermission == PolicyRead {
+			return true, false
+		} else {
+			return false, false
+		}
+	case PolicyRead:
+		// grants just read permissions
+		if requiredPermission == PolicyRead {
+			return true, false
+		} else {
+			return false, false
+		}
+	case PolicyDeny:
+		// explicit denial - do not recurse
+		return false, false
+	default:
+		// need to recurse as there was no specific policy set
+		return false, true
+	}
+}
```

Comment:
Given the importance of this function, I would recommend isolated unit testing with a table of this, too. Usually I don't test non-exported functions but if they're complex enough or important enough I recommend it. In this case, this function has a lot of branch possibilities and the branching is _really_ important. The high-level ACL tests in theory should find issues but having 100% branch coverage over this individual function I think is very important.

Reply:
Done